### ISSUE TITLE
Sign-in callback returns SyntaxError: Unexpected token 'e" is not valid JSON

### ISSUE THREAD

**Issue by iohcidnal** [workos/authkit-nextjs#96]
Using:

next: 14.2.13
@workos-inc/authkit-nextjs: ^0.11.1

After email confirmation, the sign-in callback returns:

```
⨯ SyntaxError: Unexpected token 'e" is not valid JSON
    at JSON.parse (<anonymous>)
    at GET (webpack-internal:///(rsc)/../../node_modules/@workos-inc/authkit-nextjs/dist/esm/authkit-callback-route.js:22:43)
```

Not sure what I'm doing wrong  as I followed the steps from the docs: https://workos.com/docs/user-management/nextjs/2-add-authkit-to-your-app/callback-route

**Comment by PaulAsjes** [comment]
Hey there, looks like something is going wrong when trying to parse the `state` parameter in the callback URL which should contain the path name of the page your user is trying to hit. 

Could you post the full callback URL after being redirected from AuthKit? Should look something like `http://localhost:3000/callback?code=123&state=abc`

**Comment by iohcidnal** [comment]
> Hey there, looks like something is going wrong when trying to parse the `state` parameter in the callback URL which should contain the path name of the page your user is trying to hit.
> 
> Could you post the full callback URL after being redirected from AuthKit? Should look something like `http://localhost:3000/callback?code=123&state=abc`

Thank you for looking into this, @PaulAsjes. Here is the full callback URL: http://localhost:3000/callback?code=01J91MGR26Q4S5YDVRFKWDCZJ8&state=null

Thanks again!

**Comment by PaulAsjes** [comment]
`state=null` is why that JSON is failing to parse. Could you share more of your code? Specifically how you're calling `handleAuth` and `authkitMiddleware`.

**Comment by iohcidnal** [comment]
callback/route.ts:
```
import { handleAuth } from "@workos-inc/authkit-nextjs";

export const runtime = "edge"; // needed for cloudflare pages
export const GET = handleAuth();
```

middleware.ts
```
import { authkitMiddleware } from "@workos-inc/authkit-nextjs";

export default authkitMiddleware();

export const config = { matcher: ["/"] };
```

**Comment by iohcidnal** [comment]
I removed runtime edge and Cloudflare-related configs and still got the same error.

**Comment by iohcidnal** [comment]
My setup is the same as found in: https://github.com/workos/next-authkit-example
The only difference is I'm using turborepo. So the web project is one of the packages.

**Comment by iohcidnal** [comment]
Also, where is the `state` generated and constructed from? From what I'm seeing, that is coming from the server where the invite came from, right? So, if I have an invite link of:

https://foo-bar.authkit.app/invite/?invitation_token=123

Then is the `state` generated by `foo-bar`?

**Comment by PaulAsjes** [comment]
`state` is used to keep track of what page your user was originally intending to view after they come back from AuthKit. It's set [here](https://github.com/workos/authkit-nextjs/blob/main/src/get-authorization-url.ts/#L15). 

In your case I'm uncertain why it's coming up as `null` as it looks like you only have the middleware enabled for the route `"/"`. 

Are you encountering the same issue when you use the default example app or are you only seeing it when using turborepo?

**Comment by iohcidnal** [comment]
I cloned the next-authkit-example and getting the same error:

```
http://localhost:3000/callback?code=01J91R4NZ4PDEYCD1FP0PA4G75&state=null
```
![Screenshot 2024-09-30 at 9 19 15 AM](https://github.com/user-attachments/assets/226755b7-9762-4702-af1d-4e83f2a8d4d6)

**Comment by iohcidnal** [comment]
My `NEXT_PUBLIC_WORKOS_REDIRECT_URI` in my `.env.local` is already set to `http://localhost:3000/callback`

**Comment by iohcidnal** [comment]
I also changed the authkit-example repo's WORKOS_REDIRECT_URI to NEXT_PUBLIC_WORKOS_REDIRECT_URI and set it to http://localhost:3000/callback, and now getting the same error:

```
SyntaxError: Unexpected token 'e" is not valid JSON
    at JSON.parse (<anonymous>)
    at GET (webpack-internal:///(rsc)/./node_modules/@workos-inc/authkit-nextjs/dist/esm/authkit-callback-route.js:22:43)
```

**Comment by iohcidnal** [comment]
Looking at https://github.com/workos/authkit-nextjs/blob/main/src/get-authorization-url.ts#L15, should this account for a falsy `returnPathname`, something like:

```
state: btoa(JSON.stringify({ returnPathname: returnPathname || '/' }))
```

**Comment by iohcidnal** [comment]
I tried the above suggestion, and it worked. The complete callback URI would be: http://localhost:3000/callback?code=01J91XGD66SNMPTF780QC2SNXX&state=eyJyZXR1cm5QYXRobmFtZSI6Ii8ifQ==

Thoughts?

**Comment by iohcidnal** [comment]
I'm using node 20.x and chrome.

**Comment by sestinj** [comment]
@PaulAsjes I'm running into this as well, though only after creating a new account from an invitation.

In the middle of the sign-up process (entering name, password) the URL is like `https://<MY_SUBDOMAIN>.authkit.app/sign-up?invitation_token=<INVITE_TOKEN>&invite_token=<INVITE_TOKEN>&state=null&redirect_uri=http%3A%2F%2Flocalhost%3A3000%2Fcallback&email=<EMAIL>&source=invite&authorization_session_id=`

I'm trying to determine whether I'm just unaware of the correct place to set `state`, or whether this is set automatically to `null`. If I'm missing something, I'm optimistic the solution won't be terribly demanding.

If this `null` is to be expected, then it seems that it just ought to be better handled by `authkit-callback-route`:

https://github.com/workos/authkit-nextjs/blob/20aebf56d9ba9760d16938675609e854c994288d/src/authkit-callback-route.ts#L15

The line above could for example be

```ts
const returnPathname = (state && state !== "null") ? JSON.parse(atob(state)).returnPathname : null;
```

, which would cause the default path of `"/"` to be used.

I'm more than happy to put up the PR for this if it seems reasonable. Thanks!

**Comment by iohcidnal** [comment]
@sestinj I am also creating a new account from an invitation. I tried your suggestion above, and it works when accepting the invite but errors when you sign out.

I really like workOS, but now I'm looking into alternatives because of this error.

**Comment by danielr18** [comment]
I also have the same issue when creating the account from an invitation.

**Comment by d3vhound** [comment]
Yeah creating an account from an invitation should not set state=null if its just going to throw.

**Comment by PaulAsjes** [comment]
Thanks for the updates! Knowing that this comes from invites helps narrow it down. I'm looking into it and hope to have a fix out soon.

**Comment by PaulAsjes** [comment]
The fix has been released as v0.12.2! Please let us know if you're still running into problems.

**Comment by iohcidnal** [comment]
Thank you!
